// @tenex/dynamoose-object/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dynamoose = require("dynamoose");
let hashKey = "_id";
class Dynamobject {
    constructor(model, object) {
        this.model = model;
        this.object = object;
    }
    save() {
        return new this.model(this.object).save();
    }
    get(key) {
        return this.model(this.object).get({ [hashKey]: key });
    }
}
exports.toDynamoObject = (table, object) => {
    const dynamoObject = {};
    for (const key in object) {
        const jsAttr = toJavascriptAttribute(object[key]);
        dynamoObject[key] = { type: jsAttr };
        if (key == hashKey)
            dynamoObject[key].hashKey = true;
    }
    const model = dynamoose.model(table, dynamoObject, { update: true });
    return new Dynamobject(model, object);
};
function toJavascriptAttribute(value) {
    const type = value.constructor.name;
    switch (type) {
        case "String": return String;
        case "Number": return Number;
        case "Boolean": return Boolean;
        case "Object": return Object;
        case "Array": return Array;
    }
}
